FantasyUniverse/Services/Auth/FirebaseAuthService.swift

```diff
@@ -30,97 +36,159 @@ final class FirebaseAuthService: ObservableObject {
-                                    .child(AuthFirebaseKeys.users.rawValue)
-                                    .child(uid)
-                                    .updateChildValues(values) { error, _ in
-                                        if let err = error {
-                                            promise(.failure(err))
-                                        } else {
-                                            promise(.success(()))
-                                        }
-                                    }
-                                
+                                promise(.success(uid))
                             } else {
-                                promise(.failure(NSError(domain: "Invalid User Id", code: 0, userInfo: nil)))
+                                promise(.failure(.noUser))
                             }
                         }
                     }
             }
         }
-        .receive(on: RunLoop.main)
+        .flatMap { [weak self] uid in
+            self?.storeUserInfoPublisher(
+                uid: uid,
+                firstname: credentials.firstname,
+                lastname: credentials.lastname) ??
+            Fail<Void, AuthError>(error: AuthError.noUser) .eraseToAnyPublisher()
+        }
         .eraseToAnyPublisher()
     }
     
-    func resetPasswordPublisher(with email: String) -> AnyPublisher<Void, Error> {
+    func resetPasswordPublisher(with email: String) -> AnyPublisher<Void, AuthError> {
         Deferred {
             Future { promise in
                 Auth
-                    .auth()
-                    .sendPasswordReset(withEmail: email, completion: { err in
-                        if let error = err {
-                            promise(.failure(error))
+                  .auth()
+                  .sendPasswordReset(
+                    withEmail: email,
+                    completion: { err in
+                        if err != nil {
+                            promise(.failure(.cantResetPassword))
                         } else {
                             promise(.success(()))
                         }
                     })
             }
         }
-        .receive(on: RunLoop.main)
+        .eraseToAnyPublisher()
+    }
+      
+    func uploadProfileImagePublisher(_ image: UIImage) -> AnyPublisher<Void, AuthError> {
+        Deferred {
+            Future { promise in
+                if let uid = self.currentUserUID {
+                    guard let imageData = image.jpegData(compressionQuality: 0.5) else {
+                        promise(.failure(.cantCompressImage))
+                        return
+                    }
+                    let storageRef = Storage.storage().reference()
+                    let imagesStorageRef = storageRef.child(AuthFirebaseKeys.images.rawValue).child(uid)
+                    imagesStorageRef.putData(imageData, metadata: nil) { (metadata, error) in
```

Comment:
Something a little more elegant, but probably overkill would be to have `putData`, `downloadURL` and `commitChanges` separated to their own proxy methods that return a publisher and chain them with Combine instead of a piramid of doom with multiple nested completion blocks